File: examples/safety_memory_store_quickstart.cpp
#include <rbfsafe/rbfsafe.h>

#include <filesystem>
#include <iostream>
#include <string>

namespace {

std::string digest(char value) { return std::string(64, value); }

rbfsafe::MemoryArtifactInput artifact() {
    rbfsafe::MemoryArtifactInput result;
    result.type = rbfsafe::MemoryArtifactType::SafeAtlas;
    result.deployment_id = "arm-a";
    result.robot_digest = digest('a');
    result.scene_digest = digest('c');
    result.task_id = "shelf-pick";
    result.content_digest = digest('1');
    result.locator = "artifacts/shelf-atlas";
    result.evidence = rbfsafe::EvidenceLevel::CertifiedRegion;
    result.tags = {"production", "shelf"};
    return result;
}

} // namespace

int main(int argc, char** argv) {
    using namespace rbfsafe;
    if (argc != 2) {
        std::cerr << "usage: rbfsafe_safety_memory_store_quickstart <new-store-directory>\n";
        return 2;
    }
    SafetyMemory memory;
    auto registered = memory.register_artifact(artifact());
    if (!registered) {
        std::cerr << registered.error().describe() << '\n';
        return 1;
    }
    auto store = SafetyMemoryStore::create(std::filesystem::path(argv[1]), memory);
    if (!store) {
        std::cerr << store.error().describe() << '\n';
        return 1;
    }
    const auto root = store.value().current_revision_id();
    auto transitioned = memory.transition(registered.value().id, registered.value().generation,
                                          MemoryArtifactState::Stale, "scene maintenance");
    if (!transitioned) {
        std::cerr << transitioned.error().describe() << '\n';
        return 1;
    }
    auto published = store.value().publish(memory, root);
    if (!published) {
        std::cerr << published.error().describe() << '\n';
        return 1;
    }
    auto current_memory = store.value().load_current();
    if (!current_memory) {
        std::cerr << current_memory.error().describe() << '\n';
        return 1;
    }
    std::cout << "root=" << root << '\n'
              << "current=" << published.value().id << '\n'
              << "revisions=" << store.value().revisions().size() << '\n'
              << "stale=" << current_memory.value().summary().stale << '\n';
    return 0;
}

File: examples/safety_memory_store_quickstart.py
"""Create and publish immutable revisions in a safety-memory store."""

from __future__ import annotations

import argparse
from pathlib import Path

import rbfsafe


def digest(value: str) -> str:
    return value * 64


parser = argparse.ArgumentParser()
parser.add_argument("store", type=Path, help="new safety-memory store directory")
args = parser.parse_args()

artifact = rbfsafe.MemoryArtifactInput()
artifact.type = rbfsafe.MemoryArtifactType.SAFE_ATLAS
artifact.deployment_id = "arm-a"
artifact.robot_digest = digest("a")
artifact.scene_digest = digest("c")
artifact.task_id = "shelf-pick"
artifact.content_digest = digest("1")
artifact.locator = "artifacts/shelf-atlas"
artifact.evidence = rbfsafe.EvidenceLevel.CERTIFIED_REGION
artifact.tags = ["production", "shelf"]

memory = rbfsafe.SafetyMemory()
registered = memory.register_artifact(artifact)
store = rbfsafe.SafetyMemoryStore.create(args.store, memory)
root = store.current_revision_id
memory.transition(
    registered.id,
    registered.generation,
    rbfsafe.MemoryArtifactState.STALE,
    "scene maintenance",
)
current = store.publish(memory, root)
print(f"root={root}")
print(f"current={current.id}")
print(f"revisions={len(store.revisions)}")
print(f"stale={store.load_current().summary.stale}")
